### ISSUE TITLE
Is this possible to be implemented with hybrid model (R50+ViT) ?

### ISSUE THREAD

**Issue by gwang-kim** [hila-chefer/Transformer-Explainability#8]
https://github.com/rwightman/pytorch-image-models/blob/f8463b8fa9c0490db093b36acfce71fa2363b8c3/timm/models/vision_transformer.py#L169

I can't found HybridEmbed in your implementation.

Could you suggest how to implement RelProp for HybridEmbed?

I want to use this your visualization method with this ResNetV2 implementation.
https://github.com/jeonsworld/ViT-pytorch/blob/460a162767de1722a014ed2261463dbbc01196b6/models/modeling_resnet.py#L129

**Comment by gwang-kim** [comment]
\+
```
class GroupNorm(nn.GroupNorm, RelProp):
    pass

```
Can GroupNorm be implemented with this in layers_ours.py?
or 
Require some modification for RelProp?

Thx in advance!!

**Comment by hila-chefer** [comment]
Hi @gwang-kim, thanks for your interest in our work!

In our paper, we only included pure Transformer networks, so I removed the hybrid model implementation, but I can recommend several resources for getting LRP implementation: [link to innvestigate](https://github.com/albermax/innvestigate/blob/b1084b2b5c59434060c78bb163b9bf006f5bbeb8/innvestigate/analyzer/relevance_based/relevance_rule.py#L218) which is a great resource, this link is specifically to the LRP alpha beta formulation we use in our paper with alpha=0 (i.e. only considering positive contributions), it should really help in implementation.
In addition, there's [the RAP repo](https://github.com/wjNam/Relative_Attributing_Propagation/blob/master/modules/resnet.py) which is for a paper co-authored by @shirgur, I believe you'll find implementations there for probably anything you need (notice to use relprop and not RAP_relprop, it's a different method).

I hope this helps, it should cover all you require to implement LRP, and of course, I highly encourage you to share your results once you're done, and create a PR if you wish to :)

**Comment by hila-chefer** [comment]
Hi @gwang-kim actually, coming to think about it- you can use our method without LRP for the backbone.
In order to use our method, all you need is just propagating from the last layer to all attention layers and you're done. Since the backbone is used for patch embedding, it should work with our implementation without further changes, if you wish, I can add an implementation for that given pretrained weights for the network (which I don't see in the repo currently...)

**Comment by gwang-kim** [comment]
Great @hila-chefer , thanks for your kind answer and It will be so great if you implement it!

There are official hybrid checkpoints here
https://console.cloud.google.com/storage/browser/vit_models;tab=objects?prefix=&forceOnObjectsSortingFiltering=false

And below two repos implemented pytorch version of the hybrid ViT!
https://github.com/rwightman/pytorch-image-models/blob/f8463b8fa9c0490db093b36acfce71fa2363b8c3/timm/models/vision_transformer.py#L75

https://github.com/jeonsworld/ViT-pytorch/blob/460a162767de1722a014ed2261463dbbc01196b6/models/modeling.py#L142

**Comment by hila-chefer** [comment]
@gwang-kim sure, I'll leave this issue open for now, as I'm not sure that Google's published weights are compatible with this ViT implementation, but in the future, once they'll be added to [Ross Wightman's repo](https://github.com/rwightman/pytorch-image-models) I'll add an implementation here as well.

**Comment by gwang-kim** [comment]
Great, thanks a lot!

**Comment by gwang-kim** [comment]
Hi, @hila-chefer . I wonder whether GradCam that is applied to last transformer layer is equal to  method=="last".

**Comment by hila-chefer** [comment]
@gwang-kim I refer you to [this issue](https://github.com/hila-chefer/Transformer-Explainability/issues/6) regarding Grad-CAM, since it does not require LRP, I'm using the ViT model without the LRP propagation for it... last refers to either using our method only on the last layer (ablation) or using only LRP for the last layer without gradient integration

**Comment by gwang-kim** [comment]
Thx @hila-chefer 😄
